vizro-core/hatch.toml

```diff
@@ -1,11 +1,11 @@
 [envs.all]
 
 [[envs.all.matrix]]
-python = ["3.8", "3.9", "3.10", "3.11"]
+python = ["3.8", "3.9", "3.10", "3.11", "3.12"]
 
 [envs.all.overrides]
 matrix.python.features = [
-  {value = "kedro", if = ["3.8", "3.9", "3.10"]}
+  {value = "kedro", if = ["3.8", "3.9", "3.10", "3.11"]}
```

Comment:
Kedro is now compatible with 3.11

Reply:
1. In order for the tests to run you'll also need to remove `kedro = pytest.importorskip("kedro")` also. Look at the GHA logs to see which tests are currently running in the different Python versions. This would close https://github.com/McK-Internal/vizro-internal/issues/362
2. Then ideally we would just make this `matrix.python.features = ["kedro"]` and it would run on 3.12 also
3. Just glancing at the kedro source though, I think this might raise some warning which we then convert to an error... So let me know what happens after step 2 and we can figure out what to do!

Reply:
Ah, got it!

So with Python 3.12 you do get a warning converted to an error:
![Screenshot 2023-12-14 at 18 26 19](https://github.com/mckinsey/vizro/assets/90609403/404d08fc-e66a-421a-bae7-0e19cadd4ae3)

All the tests with Python 3.11 work fine 👍 

Shall we keep it as we have right now then?

Reply:
Can we do an ignore with the `filterwarnings` thing in the right file and still run the tests on 3.12?

Reply:
Somehow it doesn't work with the Kedro warning - I've tried different combinations now and also what worked already in other tests. But on the Kedro one, it just doesn't filter it. Maybe you see something obviously wrong that I can't see right now 😄 

```
@pytest.mark.filterwarnings("ignore:Kedro is not yet fully compatible with this Python version.")
def test_datasets_from_catalog(catalog_path):
    from kedro.io import DataCatalog  # noqa: E402
    from vizro.integrations.kedro import datasets_from_catalog  # noqa: E402
    ...
```

Tried these as well:
```
@pytest.mark.filterwarnings("ignore::UserWarning")
@pytest.mark.filterwarnings("ignore:Kedro")
```

or directly inside the pyproject.toml:
```
filterwarnings = [
  "error",
  'ignore:.*utcfromtimestamp.*:DeprecationWarning',  # this works
  'ignore:.*Kedro is not yet fully compatible with this Python version.*',
  'ignore:.*Kedro is not yet fully compatible.*:UserWarning',
  'ignore:.*Kedro.*:UserWarning'
]
```

However, somehow none of them work on the Kedro warning and it doesn't get filtered out during the tests 🤔

Reply:
Yeah I think you're right here. I tried to fix it in [a1b5b88](https://github.com/mckinsey/vizro/pull/216/commits/a1b5b8807ab042c6236f3731633c9bfcd792bc87) but couldn't actually get things running locally so not sure if it will work.

If it doesn't work then please can you try adding the same `ignore` to `filterwarnings` in addition to where it is now?

And if that doesn't work and you can't figure out after spending ~20 mins on it then just put it back to how you had it in [b5a6a45](https://github.com/mckinsey/vizro/pull/216/commits/b5a6a4545a0aa0c39744a78af858b8c358197db8) where we don't do `kedro` tests in 3.12.

Reply:
If you can get it working then remember to remove `kedro = pytest.importorskip("kedro")` also 🙂 Otherwise if we are skipping kedro in 3.12 then it needs to stay there.

Reply:
Interesting that you couldn't run it locally! I actually didn't have to do any of the steps you mentioned below. I just installed Python 3.12 and then `run hatch run all.py3.12:test -vv` while `all.py3.09` is my hatch default environment. This way I was able to run all the tests for each python version locally.

Yes, I've tried what you suggested above for like 30 min yesterday but couldn't get it working :( 
The Kedro warning/error keeps popping even when adding the ignores locally + globally or only one of them. I tried all combinations yesterday 😓

Reply:
So after many attempts to get my hatch.toml right I feel like the latest commit [962e4ea](https://github.com/mckinsey/vizro/pull/216/commits/962e4ea98a12537ed5d7ad7d009d749248d5338f) should work, but for some reason it doesn't even try to install kedro on our CI for Python 3.12. Does it work for you locally with that?

Reply:
Locally I get a different error now than in CI, but it's still triggered by the Kedro warning 😅 

![Screenshot 2023-12-15 at 18 03 12](https://github.com/mckinsey/vizro/assets/90609403/d4657172-59f9-49fd-a15a-45407ab54e39)

Reply:
Ah ok, this is a step forward! If you add the same `ignore` to `filterwarnings` in the `pytest.ini_options` does the test pass?

Reply:
Ah yes!! Now it does 🥳

Reply:
Nice! 🎉 

At a glance I have no idea why CI isn't installing kedro. Maybe something wrong with our hatch.toml environment configuration but no idea what. Next things I would try would be:
* remove hatch environments locally and run the same steps to see if you can reproduce the CI behaviuor
* make the `hatch pip freeze` step into `hatch -v pip freeze` to see the verbose output in CI
* login to the CI machine and try to run it on there since debugging through just repeatedly pushing to GHA to trigger the job is slow and tedious

Reply:
I was able to reproduce the CI behaviour locally but the verbose output didn't add any new insights. After trying for way too long, I've decided to revert the changes and not run the tests for Python 3.12 😞 

So we have it as before, but at least the kedro tests are additionally running for 3.11 now.

Reply:
This is fine by me, thank you for trying it out! Just to understand for future reference though:
1. how did you reproduce the CI behaviour locally? Did just removing the hatch environment and rerunning do the same thing?
2. please could you put a brief comment in the `hatch.toml` file about why kedro is excluded here with a link to this PR?

Reply:
1) I run these commands in order - I don't think the conda environment creation was actually necessary, just did it to run the same pip install command as in CI:
```
hatch env prune
conda create -n hatch_test python=3.12
conda activate hatch_test       
pip install --upgrade hatch     
hatch run all.py3.12:pip freeze   
hatch run all.py3.12:test-unit-coverage
```

-> And then I got the same CI error that Kedro was not installed

2) Yes of course!